DelayedClipboard/DelayedClipboardRenderingExplainer.md

```diff
@@ -150,11 +154,91 @@ const clipboardItemInput = new ClipboardItem({['image/png']: blobInput1, ['text/
+
+**Problems with this approach**
+Could slow down the browser/tab close operation. The timeout is a way to mitigate this performance issue, but it could lead to other problems like the web author may not have enough time to generate the payload for the format, so the clipboard would have empty data for that format.
+
+3. Choose to only keep the formats that are cheaper to produce and remove the rest (the formats that are delay rendered) by performing another clipboard.write operation: Native Excel has some logic to clear the clipboard and write a smaller subset of the initial formats that were on the clipboard when a user decides to close the app. Web authors can do this as well.
+
+**Problems with this approach**
+This would affect the experience of the user if they are unable to paste the format that has high fidelity content and is delay rendered by the web author. Also, this has to be done during `beforeunload` event which affects performance of the site and slow down browser/tab close operation.
+
+4. Throw away all the delay rendered formats and put empty data for those formats in the clipboard.
+
+**Problems with this approach**
+It would be a bad user experience, but a web author could also choose to not populate the formats that were registered for delay rendering.
+
+5. Browsers could choose to show a confirmation dialog if there is any delay rendered formats in the clipboard: Although it is similar to option 1, here the text of the dialog explicitly mentions delay rendering of formats which is more informative to the the user.
+
+**Problems with this approach**
+This option has similar concerns as option 1, but it does affect navigations, so it slows down browser/tab shutdown behavior.
+
+More info [here](https://docs.google.com/document/d/1H6ow7RWa4MeycKP3OQBoMhLuMaJct3jSCWW6apcVu9U/edit?usp=sharing)
+
+### Proposal
+
+Proposal is a hybrid of options 2 and 3
+
+1. At least one built-in format must be written to the clipboard without delay as part of the write operation so there will be some data on the clipboard (could be a low fidelity `text/plain` format) if the browser/tab closes and the site doesn't get a chance to generate the data for the delay rendered formats within the timeout window. If the site only writes one format to the clipboard, then an exception would be thrown to prevent the write operation.
+
+2. If a delay rendered callback is already running before the page unloads, cancel the callback after a timeout period if the callback hasn't completed yet. When the callback is cancelled, set empty data to the clipboard for that delay rendered format.
+
+3. Run the callback before the `beforeunload` event is fired and if it exceeds a timeout period, set empty data to the clipboard for that delay rendered format so it doesn't cause delays in navigation.
```

Comment:
If I'm understanding this part of the proposal correctly I think it could be stated a bit more precisely:
```suggestion
3. For any delay-rendered format that still hasn't been resolved at the time that the UA would fire `beforeunload`, run its callback before the `beforeunload` event is fired. For any such callback that hasn't completed within a timeout period, set empty data to the clipboard for that delay rendered format so it doesn't cause delays in navigation.
```

Reply:
Sounds good, I added this one instead. Thanks!